### ISSUE TITLE
Segmentation fault error after upgrading to 2.2

### ISSUE THREAD

**Issue by linuxjuggler** [rethinkdb/rethinkdb#5093]
Hey @danielmewes : 
I have just finished upgrading my PI2 from 2.1.5-2 to 2.2 and not sure what is the issue or why this happened.

OS : RASPBIAN JESSIE
Installation instruction : I just followed the link on the main website http://rethinkdb.com/docs/install/raspbian/ 

here is the log.

```
2015-11-15T09:38:30.098201823 1.919098s notice: Server ready, "pi2_va0" 4e70877d-7baa-48c4-ba60-d2bec46c1099
2015-11-15T09:38:33.301065227 5.121969s error: Error in src/arch/runtime/thread_pool.cc at line 359:
2015-11-15T09:38:33.301453093 5.122353s error: Segmentation fault from reading the address 0x220067a9.
2015-11-15T09:38:33.301625386 5.122523s error: Backtrace:
2015-11-15T09:38:36.004853274 7.825774s error: Sun Nov 15 09:38:33 2015\n\n1 [0x371c40]: backtrace_t::backtrace_t() at backtrace.cc:203 (discriminator 1)\n2 [0x371fc8]: format_backtrace(bool) at basic_string.h:275\n3 [0x638ea4]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83\n4 [0x3aea7c]: linux_thread_pool_t::fatal_signal_handler(int, siginfo_t*, void*) at thread_pool.cc:359\n5 [0x76b3a180]: __default_rt_sa_restorer_v2+0 at 0x76b3a180 (/lib/arm-linux-gnueabihf/libc.so.6)\n6 [0x769244]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:70\n7 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n8 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n9 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n10 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n11 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n12 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n13 [0x76a514]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:163\n14 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n15 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149\n16 [0x76a514]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:163\n17 [0x768c48]: ql::preprocess_term_tree(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, rapidjson::RAllocator*, ql::backtrace_registry_t*) at intrusive_list.hpp:13\n18 [0x73df38]: ql::json_term_storage_t::preprocess() at term_storage.cc:291\n19 [0x74e81c]: ql::query_cache_t::create(ql::query_params_t*, signal_t*) at scoped.hpp:103\n20 [0x661580]: rdb_query_server_t::run_query(ql::query_params_t*, ql::response_t*, signal_t*) at query_server.cc:43\n21 [0x37b028]: query_server_t::handle(http_req_t const&, http_res_t*, signal_t*) at server.cc:542\n22 [0x306530]: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at basic_string.h:293\n23 [0x306530]: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at basic_string.h:293\n24 [0x311408]: http_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at basic_string.h:555\n25 [0x311cc0]: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), boost::_bi::bind_t<void, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> > > >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at mem_fn_template.hpp:280\n26 [0x39b294]: linux_nonthrowing_tcp_listener_t::handle(int) at scoped.hpp:70\n27 [0x3aaee4]: coro_t::run() at coroutines.cc:214\n28 [0x3aaeb8]: coro_t::run() at coroutines.cc:182
2015-11-15T09:38:36.005984372 7.826884s error: Exiting.
```

**Comment by danielmewes** [comment]
Thanks, this appears to be in the new code that replaced protocol buffers for internally representing queries. I'm currently compiling on a PI2 to reproduce it...

The backtrace with newlines:
```
2015-11-15T09:38:36.004853274 7.825774s error: Sun Nov 15 09:38:33 2015

1 [0x371c40]: backtrace_t::backtrace_t() at backtrace.cc:203 (discriminator 1)
2 [0x371fc8]: format_backtrace(bool) at basic_string.h:275
3 [0x638ea4]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
4 [0x3aea7c]: linux_thread_pool_t::fatal_signal_handler(int, siginfo_t*, void*) at thread_pool.cc:359
5 [0x76b3a180]: __default_rt_sa_restorer_v2+0 at 0x76b3a180 (/lib/arm-linux-gnueabihf/libc.so.6)
6 [0x769244]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:70
7 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
8 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
9 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
10 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
11 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
12 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
13 [0x76a514]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:163
14 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
15 [0x769a28]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:149
16 [0x76a514]: ql::term_walker_t::walker_frame_t::walk(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, ql::backtrace_id_t) at term_walker.cc:163
17 [0x768c48]: ql::preprocess_term_tree(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::RAllocator>*, rapidjson::RAllocator*, ql::backtrace_registry_t*) at intrusive_list.hpp:13
18 [0x73df38]: ql::json_term_storage_t::preprocess() at term_storage.cc:291
19 [0x74e81c]: ql::query_cache_t::create(ql::query_params_t*, signal_t*) at scoped.hpp:103
20 [0x661580]: rdb_query_server_t::run_query(ql::query_params_t*, ql::response_t*, signal_t*) at query_server.cc:43
21 [0x37b028]: query_server_t::handle(http_req_t const&, http_res_t*, signal_t*) at server.cc:542
22 [0x306530]: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at basic_string.h:293
23 [0x306530]: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at basic_string.h:293
24 [0x311408]: http_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at basic_string.h:555
25 [0x311cc0]: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), boost::_bi::bind_t<void, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> > > >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at mem_fn_template.hpp:280
26 [0x39b294]: linux_nonthrowing_tcp_listener_t::handle(int) at scoped.hpp:70
27 [0x3aaee4]: coro_t::run() at coroutines.cc:214
28 [0x3aaeb8]: coro_t::run() at coroutines.cc:182
```

**Comment by leetal** [comment]
Reproduction is simple: Just open the dashboard. :+1:

**Comment by danielmewes** [comment]
There's finally some progress (debugging on a PI turned out to be surprisingly tricky).
I've found a way to reproduce this on AMD64 by using a different allocator, and I think I have a rough idea what the issue is.
Still figuring out the details...

**Comment by danielmewes** [comment]
Assigning @Tryneus who is going to prepare a fix for this.

**Comment by Tryneus** [comment]
A fix is up in review 3376.

**Comment by linuxjuggler** [comment]
thank you guys @Tryneus  & @danielmewes

**Comment by Tryneus** [comment]
The fix has been approved and merged to `next` in commit 1772653, and cherry-picked into `v2.2.x` in commit d8d6cb9.  Will be in release 2.2.2.

Thanks for reporting this, @linuxjuggler, and sorry you ran into it!

**Comment by danielmewes** [comment]
@linuxjuggler @leetal This should be fixed in RethinkDB 2.2.2.

**Comment by linuxjuggler** [comment]
@danielmewes great news thanks
